core/src/main/scala/kafka/log/Log.scala

```diff
@@ -544,6 +591,65 @@ class Log(val dir: File,
   }
 
   /**
+   * Get an offset based on the given timestamp
+   * The offset returned is the offset of the first message whose timestamp is greater than or equals to the
+   * given timestamp.
+   *
+   * If no such message is found, the log end offset is returned.
+   *
+   * `NOTE:` OffsetRequest V0 does not use this method, the behavior of OffsetRequest V0 remains the same as before
+   * , i.e. it only gives back the timestamp based on the last modification time of the log segments.
+   *
+   * @param timestamp The given timestamp for offset fetching.
+   * @return The offset of the first message whose timestamp is greater than or equals to the given timestamp.
+   */
+  def fetchOffsetsByTimestamp(timestamp: Long): Long = {
+    debug(s"Searching offset for timestamp $timestamp")
+    val segsArray = logSegments.toArray
+    if (timestamp == OffsetRequest.EarliestTime)
+      return segsArray(0).baseOffset
+
+    // set the target timestamp to be Long.MaxValue if we need to find from the latest.
+    val targetTimestamp = timestamp match {
+      case OffsetRequest.LatestTime => Long.MaxValue
+      case _ => timestamp
+    }
+
+    var foundOffset: Long = -1L
+    do {
+      val targetSeg = {
+        // Get all the segments whose largest timestamp is smaller than target timestamp
+        val earlierSegs = segsArray.takeWhile(_.largestTimestamp < targetTimestamp)
+        // We need to search the first segment whose largest timestamp is greater than the target timestamp if there is one.
+        if (earlierSegs.length < segsArray.length)
+          segsArray(earlierSegs.length)
+        else
+          earlierSegs.last
+      }
+
+      // First cache the current log end offset
+      val leo = logEndOffset
+      // Use the cached log end offsets if
+      // 1. user is asking for latest messages, or,
+      // 2. we are searching on the active segment and the target timestamp is greater than the largestTimestamp
+      // after we cached the log end offset. (We have to use the cached log end offsets because it is possible that
+      // some messages with a larger timestamp are appended after we check the largest timestamp. Using log end offset
+      // after the timestamp check might skip those messages.)
+      foundOffset = {
+        if (targetTimestamp == Long.MaxValue
+          || (targetTimestamp > targetSeg.largestTimestamp && targetSeg == activeSegment))
+          leo
+        else
+          targetSeg.findOffsetByTimestamp(targetTimestamp) match {
+            case Some(timestampOffset) => timestampOffset.offset
+            case None => -1L
+          }
+      }
+    } while (foundOffset < 0)
```

Comment:
Hmm, is the while loop necessary? It seems that if foundOffset is None, we can just return baseOffset of next segment or nextOffset if the search is on the active segment. Also, if we have an empty log, do we loop forever?